### PR TITLE
average ages by Yevhen Sahach

### PR DESCRIPTION
**Pull request by eugeneSahach**

### CODE REVIEW

File: src/averageAges.js [hunk 1]
```diff
@@ -15,11 +15,16 @@
  * @return {number}
  */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with logical operators (&&, ||) or ternary operator (?:)
-  // without nesting
+  let filteredMen = [];
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: do not use `let`, use `const` instead

File: src/averageAges.js [hunk 2]
```diff
@@ -15,11 +15,16 @@
...
+
+  century === undefined
+    ? filteredMen = people.filter(elem => elem.sex === 'm')
+    : filteredMen = people
+      .filter(elem => elem.sex === 'm'
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: Your code looks intuitive and readable, but you iterate over array twice, can you do it once only?

File: src/averageAges.js [hunk 3]
```diff
@@ -15,11 +15,16 @@
...
+        && century === Math.ceil(elem.died / 100));
+
+  return (filteredMen.reduce((sum, elem) => sum + (elem.died - elem.born), 0))
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: `elem` - we know elem is person

File: src/averageAges.js [hunk 4]
```diff
@@ -34,7 +39,16 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  let filteredWomen = [];
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: the same

File: src/averageAges.js [hunk 5]
```diff
@@ -34,7 +39,16 @@ function calculateMenAverageAge(people, century) {
...
+
+  withChildren === true
+    ? filteredWomen = people
+      .filter(elem => elem.sex === 'f'
+        && people.some(el => el.mother === elem.name))
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: nested loop, try to rewrite without

File: src/averageAges.js [hunk 6]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  let arrayOfAgeDifferences = [];
+
+  const isMotherDefined = kid =>
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: do not break line just after `=>`

File: src/averageAges.js [hunk 7]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
...
+    people.find(mother => mother.name === kid.mother) !== undefined;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: what about `some` you used before

File: src/averageAges.js [hunk 8]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  let arrayOfAgeDifferences = [];
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: use const instead

File: src/averageAges.js [hunk 9]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+  const isMotherWithSonDefined = kid => kid.sex === 'm'
+    && people.find(mother => mother.name === kid.mother) !== undefined;
+
+  const ageDifference = (kid) =>
+    kid.born - people.find(mother => mother.name === kid.mother).born;
+
+  onlyWithSon === undefined
+    ? arrayOfAgeDifferences = people
+      .filter(isMotherDefined)
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: nested loops inside each method

File: src/averageAges.js [hunk 10]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
...
-  // write code here
+  let arrayOfAgeDifferences = [];
+
+  const isMotherDefined = kid =>
+    people.find(mother => mother.name === kid.mother) !== undefined;
+
+  const isMotherWithSonDefined = kid => kid.sex === 'm'
+    && people.find(mother => mother.name === kid.mother) !== undefined;
+
+  const ageDifference = (kid) =>
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: functions names should start with verbs

File: src/averageAges.js [hunk 11]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  let arrayOfAgeDifferences = [];
+
+  const isMotherDefined = kid =>
+    people.find(mother => mother.name === kid.mother) !== undefined;
+
+  const isMotherWithSonDefined = kid => kid.sex === 'm'
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: `is` - more convenient for Boolean variables, but ok, not always

File: src/averageAges.js [hunk 12]
```diff
@@ -52,11 +66,31 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+  onlyWithSon === undefined
+    ? arrayOfAgeDifferences = people
+      .filter(isMotherDefined)
+      .map(ageDifference)
+    : arrayOfAgeDifferences = people
+      .filter(isMotherWithSonDefined)
+      .map(ageDifference);
+
+  return arrayOfAgeDifferences.reduce((sum, item) => sum + item, 0)
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: the same as above

File: src/averageAges.js (hunk 1 above)
Comment (yevhen-sahach) [reply src/averageAges.js]: I can't cause few lines later I assign it new value

File: src/averageAges.js [hunk 13]
```diff
@@ -15,11 +15,27 @@
...
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with logical operators (&&, ||) or ternary operator (?:)
-  // without nesting
+  let counter = 0;
+  let sum = 0;
+
+  century === undefined
+    ? people.reduce((acc, person) => {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: Why do you need reduce? You do not use acc

File: src/averageAges.js [hunk 14]
```diff
@@ -34,7 +50,28 @@ function calculateMenAverageAge(people, century) {
...
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  let sum = 0;
+  let counter = 0;
+  const mapOfPeople = people
+    .reduce((acc, person) => ({ ...acc, [person.mother]: person }), {});
+
+  withChildren === true
+    ? people.reduce((acc, person) => {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: ^

File: src/averageAges.js [hunk 15]
```diff
@@ -52,11 +89,33 @@ function calculateWomenAverageAge(people, withChildren) {
...
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  let sum = 0;
+  let counter = 0;
+  const mapOfPeople = people
+    .reduce((acc, person) => ({ ...acc, [person.name]: person }), {});
+
+  onlyWithSon === undefined
+    ? people.reduce((acc, kid) => {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: ^

File: src/averageAges.js [hunk 16]
```diff
@@ -52,11 +89,33 @@ function calculateWomenAverageAge(people, withChildren) {
...
+      if (kid.mother !== null && mapOfPeople[kid.mother] !== undefined) {
+        sum += kid.born - mapOfPeople[kid.mother].born;
+      } else {
+        counter++;
+      }
+    }, 0)
+    : people.reduce((acc, kid) => {
+      if (kid.sex === 'm'
+        && kid.mother !== null && mapOfPeople[kid.mother] !== undefined) {
+        sum += kid.born - mapOfPeople[kid.mother].born;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: duplicated code, in other methods too

Comment (YuriiHoliuk) [reply src/averageAges.js]: But this is not very critical

File: src/averageAges.js [hunk 17]
```diff
@@ -34,7 +50,28 @@ function calculateMenAverageAge(people, century) {
...
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  let sum = 0;
+  let counter = 0;
+  const mapOfPeople = people
+    .reduce((acc, person) => ({ ...acc, [person.mother]: person }), {});
+
+  withChildren === true
+    ? people.reduce((acc, person) => {
+      if (person.sex === 'f' && mapOfPeople[person.name] !== undefined) {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: in this case `mapOfPeople[person.name] !== undefined` -> `!!mapOfPeople[person.name]`, but ok

**Pull request closed without merging**